src/src/PluginStructs/P145_data_struct.cpp

```diff
@@ -0,0 +1,598 @@
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the ppm of CO2 sensed (assuming only CO2 in the air)
+@param[in] sensorType   Sensor type identifier
+@param[in] rSensor      Sensor resistance value Rs
+@param[in] rZero        R0 Resistance zero value
+@return The ppm of CO2 in the air
+*/
+/*****************************************************************************/
+float P145_data_struct::getPPM(float rSensor)
+{
+  switch (algorithm)
+  {
+    case p145AlgA:  // MQ-135
+      return (sensordef.para * pow((rSensor/rzero), -sensordef.parb));
+      break;
+    case p145AlgB:  // Miquel5612 Exponential
+      return (sensordef.para *pow((rSensor/rzero), sensordef.parb));
+      break;
+    case p145AlgC:  // Miquel5612 Linear
+      return (pow(10.0f, (log10(rSensor/rzero)-sensordef.parb)/sensordef.para));
+      break;
+    default:
+      return (0.0f);
+      break;
+  }
+  return 0.0f;   // Default value should never be returned
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the ppm of CO2 sensed (assuming only CO2 in the air), corrected
+        for temp/hum
+@param[in] rSensor      Sensor resistance value Rs
+@param[in] temperature  The ambient air temperature
+@param[in] humidity     The relative humidity
+@return The ppm of CO2 in the air
+*/
+/*****************************************************************************/
+float P145_data_struct::getCorrectedPPM(float rSensor, float temperature, float humidity)
+{
+  switch (algorithm)
+  {
+    case p145AlgA:
+      float c;                      // Temperature & humidity correction factor
+      if (temperature < 20)
+      {
+        c = sensordef.cora * temperature * temperature - sensordef.corb * temperature + sensordef.corc - (humidity - 33.) * sensordef.cord;
+      }
+      else
+      {
+        c = sensordef.core * temperature + sensordef.corf * humidity + sensordef.corg;
+      }
+      return (sensordef.para * pow(((rSensor / c) / rzero), -sensordef.parb));
+    case p145AlgB:
+      // TODO Still missing a formula to correct for temp/hum when applying this algorithm
+      break;
+    case p145AlgC:
+      // TODO Still missing a formula to correct for temp/hum when applying this algorithm
+      break;
+    default:
+      // Do nothing, return a default value at the end
+      break;
+  }
+  return 0.0f;  // Default value should never be returned
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the analog input value
+@param[in] event   Plugin context
+@return The the filtered value of the analog input over the sampling interval
+@note   The sensor is expected to be sampled ten times per second.
+        Lowest and highest value is dropped to remove outliers.
+        All other values are averaged to a single float value.
+        This algorithm is copied from P002_ADC.
```

Comment:
I think we should turn this into some statistics class, as this code gets duplicated in more places.
The same for an IIR or FIR filter.

Reply:
I copied it from the analog input as it is a good and relatively cheap algorithm to filter the analog input signal over the full measurement period. Dropping outliers seems also be handy on the ESP. As a beginning contributer without deep knowledge of C++ I will wait for others to pick it up. For this sensor it is very helpful as the active signal range is pretty small.